File: internal/lexer/lexer.go
package lexer

import (
	"fmt"
)

var lexMap = map[rune]int{
	'<': LEFT,
	'>': RIGHT,
	'+': PLUS,
	'-': MINUS,
	'[': START_LOOP,
	']': END_LOOP,
	'.': OUT,
	',': IN,
}

type Position struct {
	Line   int
	Column int
}

type LexToken struct {
	Typ      int
	Position Position
}

func Lex(code string, filePath string) []LexToken {
	fmt.Printf("[INFO] Lexing %v\n", filePath)
	lexed := make([]LexToken, 0)
	line := 0
	column := 0
	for _, c := range code {
		// filter actual symbols from "comments"
		if val, ok := lexMap[c]; ok {
			// append actual symbols to the lexed symbols
			lexed = append(lexed, LexToken{
				Typ: val,
				Position: Position{
					Line:   line,
					Column: column,
				},
			})
		}
		if c == '\n' {
			line++
			column = 0
		} else {
			column++
		}
	}
	return lexed
}

File: src/lexing/lexer.rs
use crate::lexing::tokens::{from, Position};

use super::tokens::Token;

pub fn lex(code: &String, filepath: &String) -> Vec<Token> {
    println!("[INFO] Lexing {}", filepath);
    let mut lexed: Vec<Token> = vec![];
    let mut line = 0;
    let mut column = 0;

    for c in code.chars() {
        println!("{}", c);

        if let Some(token) = from(c, Position { column, line }) {
            lexed.push(token);
        }

        if c == '\n' {
            line += 1;
            column = 0;
        } else {
            column += 1;
        }
    }

    return lexed;
}
